### DISCUSSION TITLE
AS4RequestHandler: Different keystores for decryption and signing of response message

### DISCUSSION THREAD

**Discussion by sopgreg** [phax/phase4#138] [Q&A]
Right now, in the `AS4RequestHandler` the same `IAS4CryptoFactory` is used for decryption/signature verification as well as for signing/encrypting the response message (signal/error) message.

In the German BDEW process, different keys must be used for encryption and signature purposes. Right now, since the same crypto factory is used, the response message will be signed with the private key defined in the crypto factory, which actually is the private key used for decryption.

Here, a different private key must be used.

Is there currently a way to achieve this, or would a change be necessary in the `AS4RequestHandler` code, like two separate `IAS4CryptoFactory` instances?

**Comment by phax** [comment]
Yes, that is currently not supported. I created #139 to deal with that properly.

**Reply by phax** [reply]
It is now implemented and will be part of release 2.2.0.
